flink-ml-lib/src/main/java/org/apache/flink/ml/feature/minmaxscaler/MinMaxScaler.java

```diff
@@ -0,0 +1,165 @@
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    public static MinMaxScaler load(StreamExecutionEnvironment env, String path)
+            throws IOException {
+        return ReadWriteUtils.loadStageParam(path);
+    }
+
+    /**
+     * Generates minMax scaler model data.
+     *
+     * @param minMaxVectors Input distributed minMaxVectors.
+     * @return MinMax scaler model data.
+     */
+    private static DataStream<MinMaxScalerModelData> genModelData(
+            DataStream<Tuple2<DenseVector, DenseVector>> minMaxVectors) {
+        DataStream<MinMaxScalerModelData> modelData =
+                DataStreamUtils.mapPartition(
+                        minMaxVectors,
+                        new RichMapPartitionFunction<
+                                Tuple2<DenseVector, DenseVector>, MinMaxScalerModelData>() {
+                            @Override
+                            public void mapPartition(
+                                    Iterable<Tuple2<DenseVector, DenseVector>> dataPoints,
+                                    Collector<MinMaxScalerModelData> out) {
+                                DenseVector minVector = null;
+                                DenseVector maxVector = null;
+                                int vecSize = 0;
+                                for (Tuple2<DenseVector, DenseVector> dataPoint : dataPoints) {
+                                    if (maxVector == null) {
+                                        vecSize = dataPoint.f0.size();
+                                        maxVector = dataPoint.f1;
+                                        minVector = dataPoint.f0;
+                                    }
+                                    for (int i = 0; i < vecSize; ++i) {
+                                        minVector.values[i] =
+                                                Math.min(
+                                                        dataPoint.f0.values[i],
+                                                        minVector.values[i]);
+                                        maxVector.values[i] =
+                                                Math.max(
+                                                        dataPoint.f1.values[i],
+                                                        maxVector.values[i]);
+                                    }
+                                }
+                                out.collect(new MinMaxScalerModelData(minVector, maxVector));
+                            }
+                        });
+        modelData.getTransformation().setParallelism(1);
+        return modelData;
+    }
+
+    /**
+     * Computes max and min values of features.
+     *
+     * @param inputData Input data.
+     * @param featureCol Feature column name.
+     * @return Max and min values of features.
+     */
+    private DataStream<Tuple2<DenseVector, DenseVector>> computeMinMaxVectors(
```

Comment:
The logic of `computeMinMaxVectors` and `genModelData` seems almost the same. Is merging them into one a better implementation?

Reply:
computeMinMaxVectors execute in parallel mode, genModelData reduce distributed minMax values to one.